# Introduction
We'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

### Goals
- The aim of this project is to clean the data and analyze the included used car listings using pandas.
- We'll also become more familiar with some of the unique benefits jupyter notebook provides for pandas.

### Given: Data Set
The dataset was originally scraped and uploaded to [Kaggle](https://www.kaggle.com/orgesleka/used-cars-database/data), but it was reduced to be 50,000 rows and dirtied up a bit to give me a bigger headache (and more practice!) than usual. :-)

--------
## Step 1: Open the dataset

We'll open our data set now and make it ready to be used for analysis.

In [1]:
#global imports
import numpy as np
import pandas as pd

# read the autos.csv CSV file into pandas
autos=pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
# let's make sure we imported the data correctly!
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

----------------
## Step 2: Explore the data
Let's print out a few lines of the datasets to see what they look like:

In [3]:
# let's inspect the data closer to see what
# data types we're going to be working with...
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
# let's get the column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

### Rename column names
Hmmm... There is quite a bit of inconsistency in the naming of the columns. Let's convert the column names from camelCase to snake_case + reword some of the column names based on the data dictionary to be more descriptive. Afterward, this will make accessing the columns and rows much easier since we'll have a style convention for our header row! Let's get started:

In [5]:
# rename columns
autos.columns = autos.columns.str.strip()
autos.rename({"yearOfRegistration":"registration_year",
                   "monthOfRegistration":"registration_month",
                   "notRepairedDamage":"unrepaired_damage",
                  "dateCreated":"ad_created"},
             axis='columns', inplace = True)


# change remaining col names from camelCase to snake_case
col_names=autos.columns
def camel_to_snake(list_of_strings):
    new_cols=[]
    for name in list_of_strings:
        letters=[]
        for letter in name:
            if letter==letter.upper():
                letters.append("_")
                letters.append(letter.lower())
            else:
                letters.append(letter)
        letters="".join(letters)
        letters=letters.replace("__","_")
        new_cols.append(letters)
    return new_cols

### reassign col names
autos.columns=camel_to_snake(col_names)

In [6]:
#make sure col names are updated as expected
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Let's keep exploring our dataset to see what else needs to be cleaned and fixed.

In [7]:
# describe the data, include all cols, numeric and non-numeric
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-19 17:36:18  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### Identify other issues to fix

It looks like we'll need to spend some time on the following issues:
- Columns that have mostly one value that are candidates to be dropped:
  - `seller`
  - `offer_type`
  - `abtest`
  - `gearbox`
  - `not_repaired_damaged`
- Columns that need more investigation:
  - `month_of_registration` (min: 0)
  - `nr_of_pictures` (all calculations are 0)
  - `registration_year` (min: 1000)
- Examples of numeric data stored as text that needs to be cleaned:
  - `price`
  - `odometer`

In [8]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

Since all rows of picture counts are 0, let's remove the `nr_of_pictures` column. And since `seller` and `offer_type` only have two different values in all 50,000 rows of data, let's get rid of those two columns as well.

In [9]:
# delete the nr_of_pictures, seller, and offer_type columns
autos = autos.drop(["nr_of_pictures", "seller", "offer_type"], axis="columns")

In [10]:
# convert price and odometer col data to numeric:
# replace all non-numeric characters with ""
# change col type to int
# rename col type if needed (since we lost some detail from our data)

autos["price"] = (autos["price"]
                          .str.replace("$","")
                          .str.replace(",","")
                          .astype(int)
                          )

autos["odometer"]=(autos["odometer"]
                        .str.replace("km","")
                       )
autos.rename({"odometer":"odometer_km"}, axis="columns", inplace = True)

### Continue cleaning out columns that don't look right
Let's fix any inconsistencies in the data for the `odometer_km` and `price` columns.

In [11]:
prices=autos["price"]
odo=autos["odometer_km"]

# print the shape of unique() method to count how many unique values
print("price uniques:")
print(prices.unique().shape[0])
print("\nodometer uniques:")
print(odo.unique().shape[0])

# print describe() method for both series
print("\nprice describe:")
print(prices.describe())
print("\nodometer describe:")
print(odo.describe())

# print value_counts() for both series
print("\nprice value counts (head: 10):")
print(prices.value_counts().head(10))
print("\nodometer value counts:")
print(odo.value_counts())

price uniques:
2357

odometer uniques:
13

price describe:
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

odometer describe:
count       50000
unique         13
top       150,000
freq        32424
Name: odometer_km, dtype: object

price value counts (head: 10):
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64

odometer value counts:
150,000    32424
125,000     5170
100,000     2169
90,000      1757
80,000      1436
70,000      1230
60,000      1164
50,000      1027
5,000        967
40,000       819
30,000       789
20,000       784
10,000       264
Name: odometer_km, dtype: int64


It looks like, out of 50,000 prices, 1,421 are cars sold at \\$0. Since this represents just 2.8% of our rows, we should just delete them from our dataset. Another anomoly is that there were one or more cars sold at \\$1,000,000. We should probably check out the `price` column more in-depth to see if there are other super-expensive cars or if this \\$1,000,000 car is just an outlier.

As for the odometer, things seem to look mostly normal. More than half of our dataset includes heavily-driven cars being sold (over 150,000 miles)!

In [12]:
# get more price info on the higher-selling cars
prices.value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

Sorting by descending order on the `price` column, we can see that the prices ascend normally until they reach \\$350,000. After that, the prices start looking more falsified (e.g.: \\$12,345,678 and \\$1,234,566). So let's remove any row that shows a car sold below \\$1 or above \\$350,000.

In [13]:
autos=autos[(autos["price"]<350999) & (autos["price"]>0)]
autos["price"].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

Removing those rows got us down to only 48,565 listings! Now, the quartiles, mean, and standard deviation are a little more realistic. Let's see what else we need to fix.

### Explore the date columns more in-depth
There are five date columns. Here's what they are named, and where they came from:
- `date_crawled`: added by the crawler -- `string` type (!)
- `last_seen`: added by the crawler -- `string` type (!)
- `ad_created`: from the website -- `string` type (!)
- `registration_month`: from the website
- `registration_year`: from the website

We need to convert the first three date columns from `string` type to numerical types.

Let's start by exploring what these three columns look like:

In [14]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

We can extract the first 10 characters to represent the date:

In [15]:
autos["date_crawled"]=autos["date_crawled"].str[:10]
autos["ad_created"]=autos["ad_created"].str[:10]
autos["last_seen"]=autos["last_seen"].str[:10]

# show first five rows to ensure our new date got saved
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled  ad_created   last_seen
0   2016-03-26  2016-03-26  2016-04-06
1   2016-04-04  2016-04-04  2016-04-06
2   2016-03-26  2016-03-26  2016-04-06
3   2016-03-12  2016-03-12  2016-03-15
4   2016-04-01  2016-04-01  2016-04-01

Let's explore these columns more in-depth now that we've fixed the date columns.

In [16]:
# date_crawled column
date=autos["date_crawled"]
print("date value counts as percent of whole:")
date.value_counts(normalize=True, dropna=False).sort_index()* 100

date value counts as percent of whole:


2016-03-05    2.532688
2016-03-06    1.404304
2016-03-07    3.601359
2016-03-08    3.329558
2016-03-09    3.308967
2016-03-10    3.218367
2016-03-11    3.257490
2016-03-12    3.691959
2016-03-13    1.566972
2016-03-14    3.654896
2016-03-15    3.428395
2016-03-16    2.960980
2016-03-17    3.162772
2016-03-18    1.291053
2016-03-19    3.477813
2016-03-20    3.788737
2016-03-21    3.737259
2016-03-22    3.298672
2016-03-23    3.222485
2016-03-24    2.934212
2016-03-25    3.160712
2016-03-26    3.220426
2016-03-27    3.109235
2016-03-28    3.486050
2016-03-29    3.409863
2016-03-30    3.368681
2016-03-31    3.183363
2016-04-01    3.368681
2016-04-02    3.547823
2016-04-03    3.860805
2016-04-04    3.648718
2016-04-05    1.309585
2016-04-06    0.317101
2016-04-07    0.140019
Name: date_crawled, dtype: float64

The values in `date_crawled` are mostly uniform. It does seem like, from April 5th to the 7th, there possibly weren't that many car sale postings published because the crawl rate slowed down quite a bit during that date range.

In [17]:
# ad_created column
ad=autos["ad_created"]
print("ad value counts as percent of whole:")
ad.value_counts(normalize=True, dropna=False, sort=True).sort_index()*100

ad value counts as percent of whole:


2015-06-11    0.002059
2015-08-10    0.002059
2015-09-09    0.002059
2015-11-10    0.002059
2015-12-05    0.002059
                ...   
2016-04-03    3.885514
2016-04-04    3.685782
2016-04-05    1.181921
2016-04-06    0.325337
2016-04-07    0.125605
Name: ad_created, Length: 76, dtype: float64

The `ad_created` column is a little different from the `date_crawled` and `last_seen` columns because the dates range from June 2015 to April 2016. With the other two columns, the dates range from March 5th, 2016 to April 7th, 2016.

In [18]:
# last_seen column
seen=autos["last_seen"]
print("last seen value counts as percent of whole:")
seen.value_counts(normalize=True, dropna=False).sort_index()* 100

last seen value counts as percent of whole:


2016-03-05     0.107073
2016-03-06     0.432410
2016-03-07     0.539483
2016-03-08     0.741275
2016-03-09     0.959539
2016-03-10     1.066612
2016-03-11     1.237517
2016-03-12     2.378256
2016-03-13     0.889529
2016-03-14     1.260167
2016-03-15     1.587563
2016-03-16     1.645218
2016-03-17     2.808607
2016-03-18     0.735097
2016-03-19     1.583445
2016-03-20     2.065273
2016-03-21     2.063214
2016-03-22     2.137342
2016-03-23     1.853186
2016-03-24     1.976732
2016-03-25     1.921137
2016-03-26     1.680222
2016-03-27     1.564913
2016-03-28     2.085864
2016-03-29     2.234119
2016-03-30     2.477093
2016-03-31     2.378256
2016-04-01     2.279419
2016-04-02     2.491506
2016-04-03     2.520334
2016-04-04     2.448265
2016-04-05    12.476063
2016-04-06    22.180583
2016-04-07    13.194688
Name: last_seen, dtype: float64

The percent distributions here are mostly normal-looking, but when we get to the dates range of April 5th to April 7th, the percents are anomalistically much higher. Since this date range had lower crawl rates and lower ad creations, this date range may need to be removed.

In [19]:
# registration_year describe()
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Something's off about `registration_year` column: the max value is 9999, but humanity hasn't seen that year yet! And the min year is 1000, but cars weren't even invented back in the year 1000. Let's take a closer look at the distribution of `registration_year`.

In [20]:
autos[(autos["registration_year"]<1880) | (autos["registration_year"]>2019)]

date_crawled                                               name  price  \
453     2016-03-28                                         Armee_Jeep   9800   
4164    2016-03-29  Verkaufe_DESIGN_Streifen_/_Aufkleber_VW__Opel_...     49   
4549    2016-04-01                                         Kompressor   1600   
8012    2016-03-23                      Opel_GT_Karosserie_mit_Brief!    700   
8360    2016-03-11                           Vito_touret_119_Blue_Tec  42800   
10556   2016-04-01                                         UNFAL_Auto    450   
22316   2016-03-29              VW_Kaefer.__Zwei_zum_Preis_von_einem.   1500   
22799   2016-03-20                                  Subaru_Impreza_GT   9000   
24511   2016-03-17                        Trabant__wartburg__Ostalgie    490   
24519   2016-03-05  4x_Winterreifen_auf_Alufelge_der_naechste_Wint...    250   
25003   2016-03-06              Reo_Vorkriegs_Oldtimer_Rennwagen_1928   6500   
27578   2016-03-31                                   VW_GOLF_2_SYNCRO   1700   
27618   2016-03-29  Golf_1_75ps_5911km_Original_Automatik_einer_de...   7999   
32585   2016-04-02                                         UNFAL_Auto    450   
33950   2016-03-23  58er_karmann_ghia_lowlight_Kaefer__zum_restaur...   7999   
38076   2016-04-04                                 Mercedes_Benz_A180  18000   
42079   2016-03-17                      APE_50___deutsche_Papiere_!!!    500   
49153   2016-03-12                                        Corsa_c20xe   2500   
49283   2016-03-15                                         Citroen_HY   7750   
49910   2016-04-03  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...  22200   

        abtest vehicle_type  registration_year    gearbox  power_p_s  \
453       test          NaN               4500    manuell          0   
4164   control          NaN               5000        NaN          0   
4549      test          NaN               4100        NaN          0   
8012      test          NaN               9999        NaN          0   
8360   control          NaN               6200  automatik          0   
10556  control          NaN               1800        NaN       1800   
22316  control          NaN               1000    manuell          0   
22799     test          NaN               5000    manuell        420   
24511  control          NaN               1111        NaN          0   
24519     test          NaN               5000        NaN          0   
25003  control          NaN               8888        NaN          0   
27578     test          NaN               2800    manuell         98   
27618     test          NaN               5911  automatik         75   
32585  control          NaN               1800        NaN       1800   
33950     test          NaN               9999        NaN          0   
38076     test          NaN               9999        NaN          0   
42079  control          NaN               4800    manuell          3   
49153     test          NaN               5000        NaN          0   
49283  control          NaN               1001        NaN          0   
49910     test          NaN               9000  automatik        140   

          model odometer_km  registration_month fuel_type           brand  \
453      andere       5,000                   0       NaN            jeep   
4164       golf       5,000                  12       NaN      volkswagen   
4549        NaN       5,000                   0       NaN  sonstige_autos   
8012     andere      10,000                   0       NaN            opel   
8360       vito      10,000                   7    diesel   mercedes_benz   
10556       NaN       5,000                   2       NaN      mitsubishi   
22316    kaefer       5,000                   0    benzin      volkswagen   
22799   impreza       5,000                   6    benzin          subaru   
24511       NaN       5,000                   0       NaN         trabant   
24519    andere       5,000               

Looks like there are 20 rows with year values below 1880 and above 2019. Since 20/48565*100%=.041%, we can safely remove these rows without losing a lot of precious data. Let's get rid of anything before 1880 (cars were invented in 1885) and after 2019 (just in case any cars were registered for multiple years).

In [21]:
# delete the 20 rows with bad registration years
autos=autos.loc[(autos["registration_year"]>1879) & (autos["registration_year"]<2020)]
autos["registration_year"].value_counts(normalize=True).sort_index()*100

1910    0.010300
1927    0.002060
1929    0.002060
1931    0.002060
1934    0.004120
          ...   
2015    0.807498
2016    2.513132
2017    2.867443
2018    0.968174
2019    0.004120
Name: registration_year, Length: 81, dtype: float64

In total, there are 81 unique years in use in the `registration_year` column.

----------------
## Step 3: Aggregate data we're interested in
Let's aggregate the `brand` column to see variations across different car brands.

In [22]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [23]:
autos["brand"].value_counts(normalize=True)*100

volkswagen        21.281285
opel              10.864147
bmw               10.864147
mercedes_benz      9.578741
audi               8.585848
ford               6.966732
renault            4.789371
peugeot            2.945720
fiat               2.599650
seat               1.893089
skoda              1.606757
nissan             1.526419
mazda              1.522299
smart              1.429601
citroen            1.411062
toyota             1.258626
hyundai            0.994953
sonstige_autos     0.964054
volvo              0.904316
mini               0.861057
mitsubishi         0.817798
honda              0.799258
kia                0.710681
alfa_romeo         0.661242
porsche            0.591204
suzuki             0.589144
chevrolet          0.566485
chrysler           0.348131
dacia              0.265733
daihatsu           0.251313
jeep               0.220414
subaru             0.210114
land_rover         0.203934
saab               0.162736
daewoo             0.156556
jaguar             0

It looks like Volkswagen, BMW, Opel, and Mercedes-Benz have the largest share of the brand column. These all happen to be German car makers too. Volkswagen has almost double the next-highest car maker's (Opel) numbers.

Let's aggregate on these four brands.

Here's the process we'll use:
- Identify the unique values we want to aggregate by (9% is the benchmark we'll use to grab the top four car brands)
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each:
    - Subset the dataframe by the unique values
    - Calculate the mean of whichever column we're interested in (price and mileage)
    - Assign the val/mean to the dict as key/value pairs.

In [24]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .09].index
print(common_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz'], dtype='object')


In [25]:
autos["odometer_km"] = autos["odometer_km"].str.replace(",","")
autos["odometer_km"] = autos["odometer_km"].astype(int)

In [26]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

print(brand_mean_prices)

{'volkswagen': 5333, 'opel': 2941, 'bmw': 8261, 'mercedes_benz': 8526}


It looks like BMW cars and Mercedes Benzs have a much higher average resale asking price than VWs and Opels. This is because BMW and Mercedes are luxury brands.

Opel must be a budget brand because its mean price is nearly half of VW's average price.

This could explain why VW is a popular car brand - it's not too budget or too expensive for the average car buyer.

In [27]:
brand_mean_mileage = {}

def my_mean(x):
    return np.average(x, weights=np.ones_like(x) / x.size)

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

brand_mean_mileage

{'volkswagen': 128955, 'opel': 129452, 'bmw': 132682, 'mercedes_benz': 130848}

In [28]:
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)

prices_miles = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
prices_miles["mean_price"] = mean_prices
prices_miles

mean_mileage  mean_price
bmw                  132682        8261
mercedes_benz        130848        8526
opel                 129452        2941
volkswagen           128955        5333

It turns out that the mileage does not vary as much as the average prices do. All four car brands had very similar mileage averages. Both luxury brands have more miles on their cars, on average, than VW and Opel.

----------------
## Step 4: Next Steps
There's still a ton of data cleaning and analysis that can be done on this dataset!

Data cleaning next steps:
- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
- See if there are particular keywords in the name column that you can extract as new columns

Analysis next steps:
- Find the most common brand/model combinations
- Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.
- How much cheaper are cars with damage than their non-damaged counterparts?
